flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/source/reader/deserializer/AvroToRowDataConverters.java

```diff
@@ -0,0 +1,406 @@
+                } else {
+                    throw new UnsupportedOperationException(
+                            String.format(
+                                    "TIME type with Precision > %d is not supported!",
+                                    microsPrecision));
+                }
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+                int localTsPrecision = ((LocalZonedTimestampType) type).getPrecision();
+                return createTimeDatetimeConvertor(localTsPrecision);
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+                int tsPrecision = ((TimestampType) type).getPrecision();
+                return createTimeDatetimeConvertor(tsPrecision);
+            case CHAR:
+            case VARCHAR:
+                return avroObject -> StringData.fromString(avroObject.toString());
+            case BINARY:
+            case VARBINARY:
+                return AvroToRowDataConverters::convertToBytes;
+            case DECIMAL:
+                return createDecimalConverter((DecimalType) type);
+            case ARRAY:
+                return createArrayConverter((ArrayType) type);
+            case ROW:
+                return createRowConverter((RowType) type);
+            case MAP:
+            case MULTISET:
+                return createMapConverter(type);
+            case RAW:
+            default:
+                throw new UnsupportedOperationException("Unsupported type: " + type);
+        }
+    }
+
+    private static AvroToRowDataConverter createTimeDatetimeConvertor(int precision) {
+        if (precision <= millisPrecision) {
+            return avroObject ->
+                    AvroToRowDataConverters.convertToTimestamp(avroObject, millisPrecision);
+        } else if (precision <= microsPrecision) {
+            return avroObject ->
+                    AvroToRowDataConverters.convertToTimestamp(avroObject, microsPrecision);
+        } else {
+            String invalidPrecisionError =
+                    String.format(
+                            "Obtained TIMESTAMP/DATETIME value with precision ''. "
+                                    + "Precision > %d is not supported.",
+                            precision, microsPrecision);
+            throw new UnsupportedOperationException(invalidPrecisionError);
+        }
+    }
+
+    private static AvroToRowDataConverter createDecimalConverter(DecimalType decimalType) {
+        final int precision = decimalType.getPrecision();
+        final int scale = decimalType.getScale();
+        return avroObject -> {
+            final byte[] bytes;
+            if (avroObject instanceof GenericFixed) {
+                bytes = ((GenericFixed) avroObject).bytes();
+            } else if (avroObject instanceof ByteBuffer) {
+                ByteBuffer byteBuffer = (ByteBuffer) avroObject;
+                bytes = new byte[byteBuffer.remaining()];
+                byteBuffer.get(bytes);
+            } else {
+                bytes = (byte[]) avroObject;
```

Comment:
Assuming first two conditions fail, will this casting always succeed? I cant fully visualize how avroObject will always be a byte array.

Reply:
Fixed.